core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java

```diff
@@ -228,117 +187,92 @@ private Entry getEntry(ServiceRequestContext ctx, HttpRequest req) {
+                              EnumSet<FileServiceContentEncoding> supportedEncodings) throws IOException {
+        final MediaType contentType = MimeTypeUtil.guessFromPath(path);
         for (FileServiceContentEncoding encoding : supportedEncodings) {
-            final Entry entry = getEntry(path + encoding.extension, encoding.headerValue);
-            if (entry.lastModifiedMillis() != 0) {
-                return entry;
+            final String contentEncoding = encoding.headerValue;
+            final HttpFile file = findFile(ctx, path + encoding.extension, contentType, contentEncoding);
+            if (file != null) {
+                return file;
             }
         }
-        return getEntry(path, null);
-    }
 
-    private static final class CachedEntry implements Entry {
-
-        private final Entry entry;
-        private final int maxCacheEntrySizeBytes;
-        @Nullable
-        private HttpData cachedContent;
-        private volatile long cachedLastModifiedMillis;
-
-        CachedEntry(Entry entry, int maxCacheEntrySizeBytes) {
-            this.entry = entry;
-            this.maxCacheEntrySizeBytes = maxCacheEntrySizeBytes;
-            cachedLastModifiedMillis = entry.lastModifiedMillis();
-        }
+        return findFile(ctx, path, contentType, null);
+    }
 
-        @Override
-        public MediaType mediaType() {
-            return entry.mediaType();
+    @Nullable
+    private HttpFile findFile(ServiceRequestContext ctx, String path,
+                              @Nullable MediaType contentType,
+                              @Nullable String contentEncoding) throws IOException {
+        final HttpFile uncachedFile = config.vfs().get(path, config.clock(), contentType, contentEncoding);
+        final HttpFileAttributes uncachedAttrs = uncachedFile.readAttributes();
+        if (cache == null) {
+            return uncachedAttrs != null ? uncachedFile : null;
         }
 
-        @Nullable
-        @Override
-        public String contentEncoding() {
-            return entry.contentEncoding();
+        final PathAndEncoding pathAndEncoding = new PathAndEncoding(path, contentEncoding);
+        if (uncachedAttrs == null) {
+            // Non-existent file. Invalidate the cache just in case it existed before.
+            cache.invalidate(pathAndEncoding);
+            return null;
         }
 
-        @Override
-        public long lastModifiedMillis() {
-            final long newLastModifiedMillis = entry.lastModifiedMillis();
-            if (newLastModifiedMillis != cachedLastModifiedMillis) {
-                cachedLastModifiedMillis = newLastModifiedMillis;
-                destroyContent();
+        final AggregatedHttpFile cachedFile = cache.getIfPresent(pathAndEncoding);
+        if (cachedFile == null) {
+            // Cache miss. Cache if small enough.
+            if (uncachedAttrs.length() <= config.maxCacheEntrySizeBytes()) {
```

Comment:
nit: I'm curious that is it better if we check this earlier?

Reply:
Thanks. Fixed.